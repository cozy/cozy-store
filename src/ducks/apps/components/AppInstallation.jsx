import React, { Component } from 'react'

import { connect } from 'react-redux'
import { PropTypes } from 'react-proptypes'
import flags from 'cozy-flags'

import ReactMarkdownWrapper from 'ducks/components/ReactMarkdownWrapper'
import { ModalDescription, ModalHeader, ModalFooter } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'
import Button from 'cozy-ui/react/Button'

import PermissionsList from 'ducks/apps/components/PermissionsList'
import Partnership from 'ducks/apps/components/Partnership'
import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'
import Checkbox from 'cozy-ui/react/Checkbox'

import { getTranslatedManifestProperty } from 'lib/helpers'
import { hasPendingUpdate } from 'ducks/apps/appStatus'
import storeConfig from 'config'

import { APP_TYPE, getAppBySlug, installAppFromRegistry } from 'ducks/apps'

const shouldSkipPermissions = app =>
  flags('skip-low-permissions') &&
  app.label <= storeConfig.default.authorizedLabelLimit

const hasInstallation = app => !app.installed || hasPendingUpdate(app)

export class AppInstallation extends Component {
  constructor(props) {
    super(props)
    const { app } = props
    this.state = { isTermsAccepted: false }
    this.installApp = this.installApp.bind(this)
    if (shouldSkipPermissions(app) && hasInstallation(app)) {
      this.installApp()
    }
  }

  installApp() {
    if (!this.isInstallReady()) return // Not ready to be installed
    const { app, channel, installApp, onInstallOrUpdate, t } = this.props
    const isUpdate = app.installed
    const isChannelSwitch = !!channel
    installApp(app, channel, isUpdate).catch(error => {
      const appName = getTranslatedManifestProperty(app, 'name', t)
      Alerter.error(
        t('app_modal.install.alert.install_error', {
          message: error.message,
          name: appName
        }),
        {
          duration: 12000,
          buttonText: t('app_modal.install.alert.dismiss'),
          buttonAction: dismiss => dismiss()
        }
      )
      console.error(error)
    })
    if (isUpdate || isChannelSwitch || app.type !== APP_TYPE.KONNECTOR) {
      if (typeof onInstallOrUpdate === 'function') onInstallOrUpdate()
    }
  }

  componentDidUpdate = prevProps => {
    const { app, onKonnectorInstall } = this.props
    const justInstalled =
      !!prevProps.app && !prevProps.app.installed && app.installed
    if (justInstalled && app.type === APP_TYPE.KONNECTOR) {
      if (typeof onKonnectorInstall === 'function') onKonnectorInstall()
    }
  }

  acceptTerms = () => {
    this.setState(state => ({
      isTermsAccepted: !state.isTermsAccepted
    }))
  }

  isInstallReady = () => {
    const { app, isInstalling } = this.props
    const isCurrentAppInstalling = isInstalling === app.slug
    const isTermsReady =
      shouldSkipPermissions(app) || (!app.terms || this.state.isTermsAccepted)
    return !isCurrentAppInstalling && isTermsReady
  }

  render() {
    const {
      app,
      fetchError,
      installError,
      isFetching,
      isAppFetching,
      isInstalling,
      onCancel,
      t
    } = this.props
    const { isTermsAccepted } = this.state
    const isFetchingSomething = isFetching || isAppFetching
    const isCurrentAppInstalling = isInstalling === app.slug

    if (shouldSkipPermissions(app) && hasInstallation(app)) {
      return (
        <React.Fragment>
          <ModalDescription className="sto-install-loading-wrapper">
            <div className="sto-install-loading">
              <Spinner size="xxlarge" />
            </div>
          </ModalDescription>
        </React.Fragment>
      )
    }

    return (
      <React.Fragment>
        <ModalHeader title={t('app_modal.install.title')} />
        {isFetchingSomething ? (
          <ModalDescription className="sto-install-loading-wrapper">
            <div className="sto-install-loading">
              <Spinner size="xlarge" />
            </div>
          </ModalDescription>
        ) : (
          <ModalDescription>
            {app.partnership && <Partnership app={app} />}
            {app.permissions && <PermissionsList app={app} />}
            {fetchError && (
              <p className="u-error">
                {t('app_modal.install.message.version_error', {
                  message: fetchError.message
                })}
              </p>
            )}
          </ModalDescription>
        )}
        {!isFetchingSomething && !fetchError && (
          <ModalFooter className="sto-install-footer">
            {installError && (
              <p className="u-error">
                {t('app_modal.install.message.install_error', {
                  message: installError.message
                })}
              </p>
            )}
            {app.terms && (
              <div className="sto-install-terms">
                <Checkbox
                  className="sto-install-terms-checkbox"
                  onChange={this.acceptTerms}
                  checked={isTermsAccepted}
                  disabled={isInstalling}
                >
                  <ReactMarkdownWrapper
                    source={t('app_modal.install.terms', {
                      url: app.terms.url
                    })}
                  />
                </Checkbox>
              </div>
            )}
            <div className="sto-install-controls">
              <Button
                theme="secondary"
                onClick={onCancel}
                disabled={isCurrentAppInstalling}
                label={t('app_modal.install.cancel')}
                extension="full"
                className="u-mh-half"
              />
              <Button
                theme="primary"
                disabled={!this.isInstallReady() || isInstalling}
                busy={isCurrentAppInstalling}
                extension="full"
                onClick={this.installApp}
                label={
                  hasPendingUpdate(app)
                    ? t('app_modal.install.update')
                    : t('app_modal.install.install')
                }
                className="u-mh-half"
              />
            </div>
          </ModalFooter>
        )}
      </React.Fragment>
    )
  }
}

AppInstallation.propTypes = {
  appSlug: PropTypes.string.isRequired
}

const mapStateToProps = (state, ownProps) => ({
  app: getAppBySlug(state, ownProps.appSlug),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  fetchError: state.apps.fetchError,
  installError: state.apps.actionError
})

const mapDispatchToProps = dispatch => ({
  installApp: (app, channel, isUpdate) =>
    dispatch(installAppFromRegistry(app, channel, isUpdate))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AppInstallation))
