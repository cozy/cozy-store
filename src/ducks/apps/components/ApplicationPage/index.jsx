/* global cozy */
import React, { Component } from 'react'
import { Link } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import Button from 'cozy-ui/react/Button'
import FocusTrap from 'focus-trap-react'
import AppIcon from 'cozy-ui/react/AppIcon'

import ApplicationPageLoading from 'ducks/components/ApplicationPageLoading'
import { getTranslatedManifestProperty } from 'lib/helpers'
import Header from './Header'
import Gallery from './Gallery'
import Details from './Details'

import {
  preventBackgroundScroll,
  unpreventBackgroundScroll
} from 'lib/scrollHelpers.js'
import { getAppIconProps } from 'ducks/apps'

const MOBILE_PLATFORMS = ['ios', 'android']
const isMobilePlatform = name => MOBILE_PLATFORMS.includes(name)

const { BarCenter } = cozy.bar

export class ApplicationPage extends Component {
  constructor(props) {
    super(props)
    this.state = { displayBarIcon: false, activeTrap: true }
  }

  componentDidMount() {
    preventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.addEventListener('scroll', this.handleScroll, {
        passive: true
      })
    }
  }

  componentWillUnmount() {
    unpreventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.removeEventListener('scroll', this.handleScroll)
    }
  }

  mountTrap = () => {
    this.setState({ activeTrap: true })
  }

  unmountTrap = () => {
    this.setState({ activeTrap: false })
  }

  handleScroll = () => {
    if (
      !this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop > 100
    ) {
      this.setState(() => ({ displayBarIcon: true }))
    } else if (
      this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop < 100
    ) {
      this.setState(() => ({ displayBarIcon: false }))
    }
  }

  render() {
    const {
      t,
      parent,
      matchRoute,
      isFetching,
      fetchError,
      breakpoints = {},
      pauseFocusTrap,
      getApp,
      redirectTo
    } = this.props
    if (isFetching) return <ApplicationPageLoading />
    const app = getApp(matchRoute)
    if (!app) return redirectTo(`/${parent}`)

    const { displayBarIcon } = this.state
    const { isMobile } = breakpoints
    if (fetchError) {
      return (
        <p className="u-error">
          {t('app_modal.install.message.version_error', {
            message: fetchError.message
          })}
        </p>
      )
    }
    const { icon, slug, iconToLoad } = app
    const appName = getTranslatedManifestProperty(app, 'name', t)
    const namePrefix = getTranslatedManifestProperty(app, 'name_prefix', t)
    const appShortDesc = getTranslatedManifestProperty(
      app,
      'short_description',
      t
    )
    const appLongDesc = getTranslatedManifestProperty(
      app,
      'long_description',
      t
    )
    const appChanges = getTranslatedManifestProperty(app, 'changes', t)
    const mobileApps =
      app.platforms &&
      !!app.platforms.length &&
      app.platforms.reduce((mobilePlatforms, platformInfos) => {
        // force to be lower case
        platformInfos.type = platformInfos.type.toLowerCase()
        if (isMobilePlatform(platformInfos.type)) {
          mobilePlatforms.push(platformInfos)
        }
        return mobilePlatforms
      }, [])
    return (
      <FocusTrap
        focusTrapOptions={{
          onDeactivate: this.unmountTrap,
          clickOutsideDeactivates: true
        }}
        paused={pauseFocusTrap}
      >
        <div className="sto-modal-page-app">
          {isMobile && icon && !iconToLoad && (
            <BarCenter>
              <div className="sto-app-bar">
                <AppIcon
                  app={app}
                  className={`sto-app-bar-icon ${
                    !displayBarIcon ? 'sto-app-bar-icon--hidden' : ''
                  }`}
                  {...getAppIconProps()}
                />
              </div>
            </BarCenter>
          )}
          <div className="sto-app">
            <Button
              icon="back"
              tag={Link}
              to={`${parent}`}
              className="sto-app-back"
              label={t('app_page.back')}
              onClick={this.unmountTrap}
              subtle
            />
            <Header
              app={app}
              namePrefix={namePrefix}
              name={appName}
              description={appShortDesc}
              parent={parent}
            />
            {app.screenshots && !!app.screenshots.length && (
              <Gallery slug={slug} images={app.screenshots} />
            )}
            <Details
              app={app}
              description={appLongDesc}
              changes={appChanges}
              mobileApps={mobileApps}
              parent={parent}
            />
          </div>
        </div>
      </FocusTrap>
    )
  }
}

// translate is needed here for the lang props
export default translate()(withBreakpoints()(ApplicationPage))
