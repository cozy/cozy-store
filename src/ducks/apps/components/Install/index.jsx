import React, { Component } from 'react'
import { withRouter } from 'react-router'

import Modal from 'cozy-ui/react/Modal'

import getChannel from 'lib/getChannelFromSource'

import InstallModalContent from './InstallModalContent'
import InstallModalFooter from './InstallModalFooter'

export class Install extends Component {
  constructor(props) {
    super(props)
    this.gotoParent = this.gotoParent.bind(this)
    this.state = {
      previousChannel: props.channel ? getChannel(props.app.source) : null,
      isCanceling: false
    }
    if (typeof props.fetchApp === 'function') props.fetchApp(props.channel)
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.app) this.gotoParent()
  }

  async gotoParent() {
    const { app, parent, history, fetchApp } = this.props
    const { previousChannel } = this.state
    // fetch previous channel if channel switch canceled
    if (previousChannel && typeof fetchApp === 'function') {
      this.setState(() => ({ isCanceling: true }))
      await fetchApp(previousChannel)
      this.setState(() => ({ isCanceling: false }))
    }

    if (app && app.slug) {
      history.push(`${parent}/${app.slug}`)
    } else {
      history.push(parent)
    }
  }

  render() {
    const { app, installApp, isInstalling, channel, isAppFetching } = this.props
    const { isCanceling } = this.state
    if (!app) return null
    return (
      <div className="sto-modal--install">
        <Modal dismissAction={this.gotoParent} mobileFullscreen>
          <InstallModalContent
            app={app}
            isFetching={isAppFetching}
            isCanceling={isCanceling}
          />
          <InstallModalFooter
            app={app}
            installApp={installApp}
            isFetching={isAppFetching}
            channel={channel}
            isInstalling={isInstalling}
            isCanceling={isCanceling}
          />
        </Modal>
      </div>
    )
  }
}

export default withRouter(Install)
