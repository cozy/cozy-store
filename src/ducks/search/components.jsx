import StoreAppItem from 'ducks/apps/components/StoreAppItem'
import { dumpMatches } from 'ducks/search/utils'
import sortBy from 'lodash/sortBy'
import PropTypes from 'prop-types'
import React, { useCallback, useMemo } from 'react'

import flag from 'cozy-flags'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Magnifier from 'cozy-ui/transpiled/react/Icons/Magnifier'
import Input from 'cozy-ui/transpiled/react/Input'
import InputGroup from 'cozy-ui/transpiled/react/InputGroup'
import Label from 'cozy-ui/transpiled/react/Label'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

export const SearchField = ({ onChange, value }) => {
  const { t } = useI18n()
  const { isMobile } = useBreakpoints()
  const handleChange = useCallback(
    ev => {
      onChange(ev.target.value)
    },
    [onChange]
  )
  return (
    <div className="u-flex u-flex-items-center u-flex-grow-1">
      {!isMobile ? (
        <Label className="u-di u-mr-half" htmlFor="discover-search">
          {t('discover-search-field.label')}
        </Label>
      ) : null}
      <InputGroup
        prepend={
          isMobile ? (
            <Icon icon={Magnifier} className="u-pl-1 u-coolGrey" />
          ) : null
        }
      >
        <Input
          id="discover-search"
          placeholder={t('discover-search-field.placeholder')}
          onChange={handleChange}
          type="text"
          value={value}
        />
      </InputGroup>
    </div>
  )
}

export const SearchResults = ({ searchResults, onAppClick, disableClick }) => {
  const sortedSortResults = useMemo(() => {
    return sortBy(searchResults, result => result.score)
  }, [searchResults])
  return (
    <div className="u-mv-1 u-flex u-flex-wrap">
      {sortedSortResults.map(result => {
        const app = result.item
        const isDisableClick = disableClick?.(app)
        return flag('store.show-search-score') ? (
          <div>
            <StoreAppItem
              onClick={() => !isDisableClick && onAppClick(app.slug)}
              key={app.slug}
              app={app}
            />
            <small
              onClick={() => dumpMatches(result)}
              title={JSON.stringify(result.matches, null, 2)}
            >
              {result.score.toFixed(2)}
            </small>
          </div>
        ) : (
          <StoreAppItem
            onClick={() => !isDisableClick && onAppClick(app.slug)}
            key={app.slug}
            app={app}
          />
        )
      })}
    </div>
  )
}

SearchResults.propTypes = {
  onAppClick: PropTypes.func.isRequired,
  disableClick: PropTypes.func,
  searchResults: PropTypes.array
}
