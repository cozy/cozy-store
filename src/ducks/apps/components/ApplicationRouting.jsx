import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'

import IntentModal from 'cozy-ui/react/IntentModal'
import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'

import PermissionsModal from './PermissionsModal'
import UninstallModal from './UninstallModal'
import Install from './Install'
import ApplicationPage from './ApplicationPage'

import { APP_TYPE, REGISTRY_CHANNELS } from 'ducks/apps'

export class ApplicationRouting extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isInstallSuccess: false
    }
    this.getAppFromMatchOrSlug = this.getAppFromMatchOrSlug.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    // don't react on error
    if (nextProps.actionError) return
    // on install success
    if (this.props.isInstalling && !nextProps.isInstalling) {
      const { history, location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return history.replace(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      if (app.type === APP_TYPE.KONNECTOR) {
        return history.replace(`/${parent}/${app.slug}/configure`)
      } else {
        Alerter.success(t('app_modal.install.message.install_success'), {
          duration: 3000
        })
        return history.replace(`/${parent}/${app.slug}`)
      }
    } else if (this.props.isUninstalling && !nextProps.isUninstalling) {
      const { history, location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return history.replace(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      Alerter.success(t('app_modal.uninstall.message.success'), {
        duration: 3000
      })
      return history.replace(`/${parent}/${app.slug}`)
    }
  }

  getAppFromMatchOrSlug(match, slug) {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (match && match.params && match.params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  render() {
    const {
      fetchLatestApp,
      installApp,
      uninstallApp,
      updateApp,
      isFetching,
      isAppFetching,
      isInstalling,
      parent,
      history,
      actionError,
      fetchError
    } = this.props
    return (
      <div>
        <Route
          path={`/${parent}/:appSlug`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            return <ApplicationPage app={app} parent={parent} />
          }}
        />
        <Route
          path={`/${parent}/:appSlug/channel/:channel`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app || !app.isInRegistry) return history.replace(`/${parent}`)
            const channel = match.params.channel
            const isChannelAvailable = Object.values(
              REGISTRY_CHANNELS
            ).includes(channel)
            if (!isChannelAvailable) {
              return history.replace(`/${parent}/${app.slug}/manage`)
            }
            return (
              <Install
                installApp={app.installed ? updateApp : installApp}
                parent={`/${parent}`}
                fetchApp={chan => fetchLatestApp(app.slug, chan)}
                isAppFetching={isAppFetching}
                installError={actionError}
                fetchError={fetchError}
                app={app}
                isInstalling={isInstalling}
                channel={channel}
              />
            )
          }}
        />
        <Route
          path={`/${parent}/:appSlug/manage`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            if (app && app.installed) {
              if (!app.uninstallable)
                return history.replace(`/${parent}/${app.slug}`)
              return (
                <UninstallModal
                  uninstallApp={uninstallApp}
                  parent={`/${parent}`}
                  uninstallError={actionError}
                  app={app}
                />
              )
            } else {
              return (
                <Install
                  installApp={installApp}
                  parent={`/${parent}`}
                  installError={actionError}
                  app={app}
                  isInstalling={isInstalling}
                />
              )
            }
          }}
        />
        <Route
          path={`/${parent}/:appSlug/permissions`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            return <PermissionsModal app={app} parent={`/${parent}`} />
          }}
        />
        <Route
          path={`/${parent}/:appSlug/configure`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            const goToApp = () => history.replace(`/${parent}/${app.slug}`)
            if (app && app.installed && app.type === APP_TYPE.KONNECTOR) {
              return (
                <IntentModal
                  action="CREATE"
                  doctype="io.cozy.accounts"
                  options={{ slug: app.slug }}
                  dismissAction={goToApp}
                  onComplete={goToApp}
                  mobileFullscreen
                  overflowHidden
                  size="small"
                  height="35rem"
                />
              )
            } else {
              return goToApp()
            }
          }}
        />
      </div>
    )
  }
}

export default withRouter(translate()(ApplicationRouting))
