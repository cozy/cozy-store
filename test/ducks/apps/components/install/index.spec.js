'use strict'

/* eslint-env jest */

import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { Install } from 'ducks/apps/components/Install'
import getChannel from 'lib/getChannelFromSource'

import mockApp from '../../_mockPhotosRegistryVersion'

Enzyme.configure({ adapter: new Adapter() })

const appManifest = mockApp.manifest

const getAppProps = (installed, related, transparency = false) => {
  return {
    lang: 'en',
    app: Object.assign({}, appManifest, {
      installed,
      icon: 'https://mockcozy.cc/registry/photos/icon',
      related
    }),
    parent: '/myapps',
    location: {
      pathname: transparency
        ? 'myapps/photos/permissions/transparency'
        : 'myapps/photos/permissions'
    },
    history: {
      replace: jest.fn()
    },
    match: {
      match: !transparency,
      url: 'myapps/photos/permissions'
    }
  }
}

describe('Install component', () => {
  it('should be rendered correctly with provided app', () => {
    const props = getAppProps()
    const component = shallow(<Install {...props} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly if this is the transparency modal', () => {
    const props = getAppProps(null, null, true)
    const component = shallow(<Install {...props} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should not break the permissions part if no permissions property found in manifest', () => {
    const props = getAppProps(null, null)
    delete props.app.permissions
    const component = shallow(<Install {...props} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should correctly cancel and use back the correct previous channel when installing from a new registry channel', async () => {
    const props = getAppProps(null, null)
    props.channel = 'beta'
    props.fetchApp = jest.fn()
    const wrapper = shallow(<Install {...props} />)
    // fetchApp is called in the constructor
    expect(props.fetchApp.mock.calls.length).toBe(1)
    // cancel the install from new channel and use back the previous channel
    const instance = wrapper.instance()
    instance.setState = jest.fn()
    await instance.gotoParent()
    expect(props.fetchApp.mock.calls.length).toBe(2)
    // previous channel from mockApp is stable here
    expect(props.fetchApp.mock.calls[1][0]).toBe(getChannel(props.app.source))
    // should change correctly the state during the fetching
    expect(instance.setState.mock.calls.length).toBe(2)
    expect(instance.setState.mock.calls[0][0]().isCanceling).toBe(true)
    expect(instance.setState.mock.calls[1][0]().isCanceling).toBe(false)
  })

  it('should go to parent if no app provided', () => {
    const props = getAppProps(null, null)
    props.app = null
    const component = shallow(<Install {...props} />)
    expect(component.type()).toBe(null)
    // should also goToParent on update if still no app provided
    component.setProps(props)
    expect(props.history.replace.mock.calls.length).toBe(1)
    expect(props.history.replace.mock.calls[0][0]).toBe(props.parent)
  })

  it('should not gotoparent if app provided on the update', () => {
    const props = getAppProps(null, null)
    const component = shallow(<Install {...props} />)
    expect(component.type()).not.toBe(null)
    component.setProps(props) // simulate props udpate
    expect(component.type()).not.toBe(null)
    // goToParent with history.replace should not be called
    expect(props.history.replace.mock.calls.length).toBe(0)
  })

  it('should gotoparent if no app provided on the update', () => {
    const props = getAppProps(null, null)
    const appSlug = props.app.slug
    const component = shallow(<Install {...props} />)
    const propsWithoutApp = Object.assign({}, props, {
      app: null
    })
    expect(component.type()).not.toBe(null)
    component.setProps(propsWithoutApp)
    expect(component.type()).toBe(null)
    // goToParent should be called once to go to the parent view
    expect(props.history.replace.mock.calls.length).toBe(1)
    expect(props.history.replace.mock.calls[0][0]).toBe(
      `${props.parent}/${appSlug}`
    )
  })
})
