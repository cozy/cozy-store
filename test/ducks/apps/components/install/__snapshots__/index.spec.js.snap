// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Install component should be rendered correctly if this is the transparency modal 1`] = `<withRouter(Wrapper) />`;

exports[`Install component should be rendered correctly with provided app 1`] = `
<div
  className="sto-modal--install"
>
  <Unknown
    closable={true}
    dismissAction={[Function]}
    mobileFullscreen={true}
    overflowHidden={false}
    primaryType="regular"
    secondaryType="secondary"
    size="small"
  >
    <Wrapper
      app={
        Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/icon",
          "installed": undefined,
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "permissions": Object {
            "albums": Object {
              "description": "Required to manage photos albums",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.photos.albums",
            },
            "apps": Object {
              "description": "Required by the cozy-bar to display the icons of the apps",
              "type": "io.cozy.apps",
              "verbs": Array [
                "GET",
              ],
            },
            "contacts": Object {
              "description": "Required to to share photos with your contacts",
              "methods": Array [
                "GET",
                "POST",
              ],
              "type": "io.cozy.contacts",
            },
            "files": Object {
              "description": "Required for photo access",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.files",
            },
            "settings": Object {
              "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
              "type": "io.cozy.settings",
              "verbs": Array [
                "GET",
              ],
            },
          },
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "related": undefined,
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        }
      }
      isCanceling={false}
      isFetching={undefined}
    />
    <Wrapper
      app={
        Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/icon",
          "installed": undefined,
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "permissions": Object {
            "albums": Object {
              "description": "Required to manage photos albums",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.photos.albums",
            },
            "apps": Object {
              "description": "Required by the cozy-bar to display the icons of the apps",
              "type": "io.cozy.apps",
              "verbs": Array [
                "GET",
              ],
            },
            "contacts": Object {
              "description": "Required to to share photos with your contacts",
              "methods": Array [
                "GET",
                "POST",
              ],
              "type": "io.cozy.contacts",
            },
            "files": Object {
              "description": "Required for photo access",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.files",
            },
            "settings": Object {
              "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
              "type": "io.cozy.settings",
              "verbs": Array [
                "GET",
              ],
            },
          },
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "related": undefined,
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        }
      }
      channel={undefined}
      installApp={undefined}
      isCanceling={false}
      isFetching={undefined}
      isInstalling={undefined}
    />
  </Unknown>
</div>
`;

exports[`Install component should not break the permissions part if no permissions property found in manifest 1`] = `
<div
  className="sto-modal--install"
>
  <Unknown
    closable={true}
    dismissAction={[Function]}
    mobileFullscreen={true}
    overflowHidden={false}
    primaryType="regular"
    secondaryType="secondary"
    size="small"
  >
    <Wrapper
      app={
        Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/icon",
          "installed": null,
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "related": null,
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        }
      }
      isCanceling={false}
      isFetching={undefined}
    />
    <Wrapper
      app={
        Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/icon",
          "installed": null,
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "related": null,
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        }
      }
      channel={undefined}
      installApp={undefined}
      isCanceling={false}
      isFetching={undefined}
      isInstalling={undefined}
    />
  </Unknown>
</div>
`;
